File: LeetCode/Java Solutions/Range Sum of BST/range.java
import java.util.PriorityQueue;

class Solution {
    public int rangeSum(int[] nums, int n, int left, int right) {
        final int MOD = 1_000_000_007;
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        
        // Initialize min-heap with [sum, index, end] for each number
        for (int i = 0; i < n; i++) {
            pq.offer(new int[]{nums[i], i, i});
        }
        
        int result = 0;
        for (int i = 1; i <= right; i++) {
            int[] curr = pq.poll();
            int currSum = curr[0], start = curr[1], end = curr[2];
            
            if (i >= left) {
                result = (result + currSum) % MOD;
            }
            
            if (end + 1 < n) {
                int nextSum = (int)(((long)currSum + nums[end + 1]) % MOD);
                pq.offer(new int[]{nextSum, start, end + 1});
            }
        }
        
        return result;
    }
}

File: LeetCode/Python Solutions/Range Sum of Sorted Subarray Sums/range.py
import heapq

class Solution:
    def rangeSum(self, nums: List[int], n: int, left: int, right: int) -> int:
        MOD = 10**9 + 7
        
        # Initialize min-heap with (sum, index, end) for each number
        heap = [(num, i, i) for i, num in enumerate(nums)]
        heapq.heapify(heap)
        
        result = 0
        for i in range(1, right + 1):
            curr_sum, start, end = heapq.heappop(heap)
            
            if i >= left:
                result = (result + curr_sum) % MOD
            
            if end + 1 < n:
                next_sum = curr_sum + nums[end + 1]
                heapq.heappush(heap, (next_sum, start, end + 1))
        
        return result

